DDF 1609 integration tests for IdP server and client

spearskw opened the pull request:
@ahoffer @brendan-hofmann @AzGoalie @pklinef @stustison

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/326)
<!-- Reviewable:end -->

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        // Create a validator from the metadata schema so that we can validate the metadata
+        // TODO: Maybe don't go out to the internet for this? The schema itself is going to go out to the internet anyway though.
+        URL schemaFile = new URL(
+                "http://docs.oasis-open.org/security/saml/v2.0/saml-schema-metadata-2.0.xsd");
```
Should probably save this xsd in the resources directory and load it in from there

spearskw replied:
That one XSD depends on several others, should I put them all in the resources directory?

tbatie replied:
I believe so

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        getServiceManager().startFeature(true, "security-idp", "search-ui-app");
+
+        // Create a validator from the metadata schema so that we can validate the metadata
+        // TODO: Maybe don't go out to the internet for this? The schema itself is going to go out to the internet anyway though.
```
Has this gone through checkstyle? I thought it prohibited TODO's

AzGoalie commented:
Lots of TODO's that should either be addressed or removed.
<!-- Reviewable comment -K2nAMu3VQ5ceKHaCv4F:-1606561366 -->

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 126 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBK7DUqKePejkPDDY)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L126)):</sup>
To find the title with xpath: //title[contains(text(),Redirect)]
<!-- Reviewable comment -K2nBK7EuRknKH2xneM8:-391556881 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 134 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBTSA0OpzpqZZzyud)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L134)):</sup>
Definitely log a warning here
<!-- Reviewable comment -K2nBTSA0OpzpqZZzyue:-342327379 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 179 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBXcosXijks7XjQQN)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L179)):</sup>
If this should be a 401, why not change it?
<!-- Reviewable comment -K2nBXcosXijks7XjQQO:-2125099073 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+            matcher.find();
+            fullUrl = matcher.group(1);
+        } else {
+            // TODO: this is unexpected. Log an error?
```
Should handle this else. Probably want to fail the test here since it cannot be completed at this stage

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+                .when().get(parseUrl(searchResponse) + "/sso");
+    }
+
+    private void checkSaml(Map<String, String> samlParams) throws Exception {
```
Could be a more descriptive in this method name

spearskw replied:
Do you have something in mind? validateSamlResponse maybe? That still seems a bit generic.

tbatie replied:
validateSamlResponse would work in my opinion

pklinef commented:
Make sure to clean up your commits.

harrison-tarr commented:
Your comments in the test methods explaining the flow are great! Makes it much easier to follow.
<!-- Reviewable comment -K2nYhaG8_XDoXXeQAqP:-674370524 -->

---

Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 179 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBXcosXijks7XjQQN)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L179)):</sup>
From what I remember hearing, the IdP is returning the wrong code. I agree though, I think this should be fixed now so it's not forgotten.
<!-- Reviewable comment -K2nV_BX3E7LIKvgVyTX:-552094917 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 120 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nWxyrwUv8h0YchQ8q)** ([raw file](https://github.com/codice/ddf/blob/8ee564046013232f4860432d91ac49715c047a2d/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L120)):</sup>
Would it make more sense to make this url initialized as null or just declared and not initialized? It seems right now if it hits the `else` statement, it would return an empty string. I bet this could cause quiet failures that would be harder to find the cause of than a `null` return value.
<!-- Reviewable comment -K2nWxyrwUv8h0YchQ8r:528233658 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

ahoffer commented:
We left  //TODO comments in the code in places we wanted input from reviewers. The TODOs generally pose a question.  If you have an opinion on how (or if) something should be done, leave it in the comments. When Kevin, Nate or I implement related changes, we will remove the associated //TODO.
<!-- Reviewable comment -K2qps3dpnGat8r0wkkf:-392017958 -->

---

Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

ahoffer commented:
Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 85 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2n6A0JUSU26GQB73Yr-r1-85)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L85)):</sup>
We left //TODO comments in the code in places we wanted input from reviewers. The TODOs generally pose a question. If you have an opinion on how (or if) something should be done, leave it in the comments. When Kevin, Nate or I implement related changes, we will remove the associated //TODO.
<!-- Reviewable comment -K2qrAMNlUc6JYwBObkK:-490799898 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 134 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2n6A0JUSU26GQB73Yr-r1-134)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L134)):</sup>
Agreed. Junit has a method `fail()` just for this. It can take a message as a parameter. E.g.
`fail("Unable to extract the redirect URL from the HTTP response. Text does not match regular expression.")`
<!-- Reviewable comment -K2qsKLPfaSDeECvVf-R:1674383755 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 179 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBXcosXijks7XjQQN)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L179)):</sup>
I'd rather have Scott or Phil make the call about what it should return; I don't want to modify the IdP implementation as a part of this PR. 

After discussion with Brendan, we decided to leave it expecting a 500 and leave the //TODO comment in place. When the IdP starts returning 401, the test will fail. That will be the cue to change the test to expect a 401 and to remove the //TODO.
<!-- Reviewable comment -K2qyuMF_d7ZW9AB77es:865293462 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 120 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nWxyrwUv8h0YchQ8q)** ([raw file](https://github.com/codice/ddf/blob/8ee564046013232f4860432d91ac49715c047a2d/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L120)):</sup>
Uninitialized would be fine. 
Also, declaring it where it is assigned would be fine:  `String fullUrl = response.header("Location");`
<!-- Reviewable comment -K2qrdRBYFqXm4aJ--la:1386315240 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

ahoffer commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

tbatie commented on distribution/sdk/sample-rest-endpoint/pom.xml:
```diff
@@ -0,0 +1,64 @@
+        <dependency>
+            <groupId>ddf.security.core</groupId>
+            <artifactId>security-core-api</artifactId>
+            <version>2.9.0-SNAPSHOT</version>
```
Should use a property for this version

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        Configuration serverConfig = getAdminConfig()
+                .getConfiguration("org.codice.ddf.security.idp.server.IdpEndpoint", null);
+        clientConfig.update(clientSettings);
+        serverConfig.update(serverSettings);
```
Should really wait for the configuration change to propagate. We have the new `wait` logic that you could use or this older style here - https://github.com/codice/ddf/blob/master/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/SecurityPolicyConfigurator.java#L80-L81

lessarderic replied:
We should use the new `WaitCondition` moving forward and deprecate `SynchronizedConfiguration` and all the other ways of waiting we have. That'll make the code more consistent and easier to understand.

coyotesqrl replied:
Agreed. I just looked for a quick example to include and didn't see a config change using `WaitCondition`. I should have spent more than two minutes on that.

lessarderic replied:
Don't feel bad, I didn't either :-)

Here's an example: https://github.com/codice/ddf/blob/de09be5c3e679ab9a813ca86cd6b767725a70c4e/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java#L721.

coyotesqrl replied:
:+1:

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        if (url.contains("?")) {
+            url = url.substring(0, url.indexOf('?'));
+        }
+        return url;
```
could return getRedirectUrl(response).split("[?]")[0]; instead

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+
+        // Pass our credentials to the IDP, it should redirect us to the Assertion Consumer Service.
+        // The redirect is currently done via javascript and not an HTTP redirect.
+        Response IDPResponse = given().auth().preemptive().basic("admin", "admin")
```
Since nothing is being done with the variable IDPResponse, it can be removed

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+    @Test
+    public void testIdpAuth() throws Exception {
+
+        Response initialWhoamiResponse = given().
```
Can remove initialWhoamiResponse variable; it's unused

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        Response initialWhoamiResponse = given().
+                expect().
+                statusCode(200).
+                body(not(containsString("admin"))).
```
Should use an xpath and a positive instead of a not test since you know what you are expecting.

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+                get(parseUrl(IDPResponse));
+
+        // Access search again, but now as an authenticated user.
+        Response authnSearchResponse = given().
```
remove unused var

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+                get(parseUrl(ACSResponse));
+
+        // Make sure we are logged in as admin.
+        Response whoamiResponse = given().
```
remove unused var

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+
+                expect().
+                statusCode(200).
+                body(containsString("admin")).
```
Should use an xpath.

tbatie commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java:
```diff
@@ -0,0 +1,261 @@
+        // Request search service
+        Response searchResponse = given().
+                redirects().follow(false).
+
```
Please reformat the extra blank lines and periods should be at the front of line of next method

AzGoalie commented:
Did a full build, all tests passed!

shaundmorris commented:
@codice/security I assume we are good to merge this PR?

spearskw commented:
Review status: 3 of 12 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 179 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2nBXcosXijks7XjQQN)** ([raw file](https://github.com/codice/ddf/blob/ecb640d59109f56eb8e9b2ca3fcff15acbcb5afa/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L179)):</sup>
I updated the comment to note that we're doing an AJAX call and that it's sufficient to just check that we don't get back a 200. 
<!-- Reviewable comment -K3H5AkcWlEopM4QAo7y:199672822 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 114 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2n6A0JUSU26GQB73Yr-r2-114)** ([raw file](https://github.com/codice/ddf/blob/8ee564046013232f4860432d91ac49715c047a2d/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L114)):</sup>
Nate fixed it using the example.
<!-- Reviewable comment -K3H6NnxBUa5bMMMCn-i:531905729 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 202 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2n6A0JUSU26GQB73Yr-r2-202)** ([raw file](https://github.com/codice/ddf/blob/8ee564046013232f4860432d91ac49715c047a2d/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L202)):</sup>
I just want to make sure I'm not admin, I could be guest or anonymous or something.
<!-- Reviewable comment -K3H69g1y_lY2tKl7laV:-1562709903 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java, line 255 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/326#-K2n6A0JUSU26GQB73Yr-r2-255)** ([raw file](https://github.com/codice/ddf/blob/8ee564046013232f4860432d91ac49715c047a2d/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSingleSignOn.java#L255)):</sup>
After discussing, decided to still use containsString because it is cleaner than trying to use xpath.
<!-- Reviewable comment -K3H5ydZ2kvEgex8wzTS:-1445957101 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

spearskw commented:
I've addressed all the comments people had offline, I may need to get them to confirm that though.
<!-- Reviewable comment -K3H8TAOLwea06zVewZY:-1647403322 -->

---

Review status: 3 of 12 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Reviewed 1 of 2 files at r7, 5 of 7 files at r8.
Review status: 9 of 12 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
:thumbsup: 
<!-- Reviewable comment -K3HCbVbilUdWwjYfA2Y:262468467 -->

---

Review status: 9 of 12 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

tbatie commented:
:+1: lgmt, just waiting on the restarted bamboo build

ahoffer commented:
Reviewed 3 of 6 files at r1, 1 of 2 files at r4, 1 of 2 files at r7, 6 of 7 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

ahoffer commented:
Reviewed 1 of 7 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/326)
<!-- Sent from Reviewable.io -->

shaundmorris commented:
@codice/security hopefully last call for any other comments?

shaundmorris commented:
Verified @spearskw Codice ICLA was accepted (new contributor)
